libs/autotagging-commons/src/main/java/org/opensearch/rule/InMemoryRuleProcessingService.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.rule;
+
+import org.opensearch.autotagging.Attribute;
+import org.opensearch.autotagging.FeatureType;
+import org.opensearch.autotagging.Rule;
+import org.opensearch.rule.attribute_extractor.AttributeExtractor;
+import org.opensearch.rule.storage.AttributeValueStore;
+import org.opensearch.rule.storage.AttributeValueStoreFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * This class is responsible for managing in-memory view of Rules and Find matching Rule for the request
+ * Each auto-tagging feature should use a separate instance of this class as this avoid potential concurrency overhead
+ * in case of dynamic updates and attribute sharing scenarios
+ */
+public class InMemoryRuleProcessingService {
+
+    private final AttributeValueStoreFactory attributeValueStoreFactory;
+
+    /**
+     *  Constrcutor
+     * @param featureType
+     * @param attributeValueStoreSupplier
+     */
+    public InMemoryRuleProcessingService(
+        FeatureType featureType,
+        Supplier<AttributeValueStore<String, String>> attributeValueStoreSupplier
+    ) {
+        attributeValueStoreFactory = new AttributeValueStoreFactory(featureType, attributeValueStoreSupplier);
+    }
+
+    /**
+     * Adds the rule to in-memory view
+     * @param rule to be added
+     */
+    public synchronized void add(final Rule rule) {
+        new AddRuleOperation(rule, attributeValueStoreFactory).perform();
+    }
+
+    /**
+     * Removes the rule from in-memory view
+     * @param rule to be removed
+     */
+    public synchronized void remove(final Rule rule) {
+        new DeleteRuleOperation(rule, attributeValueStoreFactory).perform();
+    }
+
+    /**
+     * Evaluates the label for the current request. It finds the matches for each attribute value and then it is an
+     * intersection of all the matches
+     * @param attributeExtractors list of extractors which are used to get the attribute values to find the
+     *                           matching rule
+     * @return a label if there is unique label otherwise empty
+     */
+    public Optional<String> evaluateLabel(List<AttributeExtractor<String>> attributeExtractors) {
+        assert attributeValueStoreFactory != null;
+        Optional<String> result = Optional.empty();
+        for (AttributeExtractor<String> attributeExtractor : attributeExtractors) {
+            AttributeValueStore<String, String> valueStore = attributeValueStoreFactory.getAttributeValueStore(
+                attributeExtractor.getAttribute()
+            );
```

Comment:
I am wondering if there can be synchronization issue when label is being evaluated and label is added/removed. And, we cannot synchronize this method for obvious reasons.

Reply:
From the consistency point of view there can be in rare scenarios and the frequency of such occurrences will be rare. But from application failure point of view there should not be an issue